packages/office-ui-fabric-react/src/components/SpinButton/SpinButton.tsx

```diff
@@ -0,0 +1,404 @@
+  }
+
+  /**
+   * Increment function to use if one is not passed in
+   */
+  private _defaultOnIncrement = (value: string) => {
+    let newValue = Math.min(+value + this.props.step, this.props.max);
+    return String(newValue);
+  }
+
+  /**
+   * Increment function to use if one is not passed in
+   */
+  private _defaultOnDecrement = (value: string) => {
+    let newValue = Math.max(+value - this.props.step, this.props.min);
+    return String(newValue);
+  }
+
+  /**
+   * Set the label direction with the gap on the correct side
+   */
+  private _labelDirectionHelper(): any {
+    let style: any = {};
+
+    switch (this.props.labelPosition) {
+      case Position.start:
+        style = { float: 'left', marginRight: this.props.labelGapSpace + 'px' };
+        break;
+      case Position.end:
+        style = { float: 'right', marginLeft: this.props.labelGapSpace + 'px' };
+        break;
+      case Position.top:
+        style = { marginBottom: this.props.labelGapSpace + 'px' };
+        break;
+      case Position.bottom:
+        style = { marginTop: this.props.labelGapSpace + 'px' };
+    }
+
+    return style;
+  }
+
+  private _onChange() {
+    /**
+     * A noop input change handler.
+     * https://github.com/facebook/react/issues/7027.
+     * Using the native onInput handler fixes the issue but onChange
+     * still need to be wired to avoid React console errors
+     * TODO: Check if issue is resolved when React 16 is available.
+     */
+  }
+
+  /**
+   * This is used when validating text entry
+   * in the input (not when changed via the buttons)
+   * @param event - the event that fired
+   */
+  private _validate(event: React.FocusEvent<HTMLInputElement>) {
+    const element: HTMLInputElement = event.target as HTMLInputElement;
+    const value: string = element.value;
+    if (this.state.value) {
+      const newValue = this._onValidate(value);
+      this._lastValidValue = newValue;
+      this.setState({ value: newValue });
+    }
+  }
+
+  /**
+   * The method is needed to ensure we are updating the actual input value.
+   * without this our value will never change (and validation will not have the correct number)
+   * @param event - the event that was fired
+   */
+  private _onInputChange(event: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>): void {
+    const element: HTMLInputElement = event.target as HTMLInputElement;
+    const value: string = element.value;
+
+    this.setState({
+      value: value,
+    });
+  }
+
+  /**
+   * Update the value with the given stepFunction
+   * @param shouldSpin - should we fire off another updateValue when we are done here? This should be true
+   * when spinning in response to a mouseDown
+   * @param stepFunction - function to use to step by
+   */
+  private _updateValue(shouldSpin: boolean, stepFunction: (string) => string) {
+    const newValue = stepFunction(this.state.value);
+    this._lastValidValue = newValue;
+    this.setState({ value: newValue });
```

Comment:
Also relevant here: https://facebook.github.io/react/docs/state-and-lifecycle.html#state-updates-may-be-asynchronous